File: salesos/backend/app/application/dashboard/dto/widget_contract.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.application.dashboard.dto.widget_state import WidgetStatus


class WidgetAction(BaseModel):
    id: str
    label: str
    type: str  # 'refresh' | 'navigate' | 'dismiss' | 'custom'
    payload: dict | None = None


class DashboardWidget(BaseModel):
    id: str
    title: str
    status: WidgetStatus = WidgetStatus.loading
    lastUpdated: datetime | None = None
    data: dict | None = None
    actions: list[WidgetAction] = Field(default_factory=list)

File: salesos/frontend/src/application/dashboard/widget.contract.ts
export type WidgetStatus = 'ready' | 'loading' | 'degraded' | 'error'

export interface WidgetAction {
  id: string
  label: string
  type: 'refresh' | 'navigate' | 'dismiss' | 'custom'
  payload?: Record<string, unknown>
}

export interface DashboardWidget<T> {
  id: string
  title: string
  status: WidgetStatus
  lastUpdated: string | null
  data: T | null
  actions: WidgetAction[]
}
